post: rendering perf 1 (#114)
SoYoung210/SOSO: content/blog/web/browser-rendering-performance.md

---
title: '렌더링 성능 개선(1) — 렌더링 과정 이해하기'
date: 2023-02-05 08:00:09
category: web
thumbnail: './images/browser-rendering-performance/thumbnail.png'
---

## 들어가며

최근 개인 프로젝트의 [렌더링 성능을 개선](https://twitter.com/th096/status/1613759830703112194)하며 브라우저 렌더링 과정에 대해 정리하고 정확히 무엇이 성능 저하의 원인인지, 어떻게 개선될 수 있었는지 정리해봤다.

## 렌더링 파이프라인

> 렌더링 성능을 개선하려면 먼저 렌더링 과정과 이에 관여하는 요소들에 대해 알아야 한다. 이 글에서 다룰 내용은 Blink엔진 기준이며, [RenderingNG](https://developer.chrome.com/articles/renderingng-architecture/) 글과 [2020년 BlinkOn발표 Life of a Pixel](https://www.youtube.com/watch?v=K2QHdgAKP-s)내용을 참고했다. 따라서 Blink엔진을 사용하지 않는 다른 브라우저에서는 내용상 차이가 있을 수 있다.

Author: ```suggestion
렌더링 성능을 개선하려면 먼저 렌더링 과정과 이에 관여하는 요소을 알아야 한다. 이 글에서 다룰 내용은 Blink엔진 기준이며, [RenderingNG](https://developer.chrome.com/articles/renderingng-architecture/) 글과 [2020년 BlinkOn발표 Life of a Pixel](https://www.youtube.com/watch?v=K2QHdgAKP-s)내용을 참고했다. 따라서 Blink엔진을 사용하지 않는 다른 브라우저에서는 차이가 있을 수 있다.
```

![렌더링_메인_플로우](./images/browser-rendering-performance/메인플로우.png)

### 1. Parsing

![parsing](./images/browser-rendering-performance/parsing.png)

가장 첫 단계로 메인 스레드에서 HTML을 브라우저가 해석할 수 있는 자료구조인 DOM Tree로 변환하는 작업이 진행된다. 이 과정은 HTML 파서가 `<link>` 또는 `async`, `defer` 가 없는 `<script>` 와 같은 블로킹 리소스를 만나기 전까지 진행된다.

CSS 파일의 경우 [스타일링이 적용되지 않는 콘텐츠가 잠깐 뜨는 현상(Flash of unstyled content, AKA FOUC)](https://en.wikipedia.org/wiki/Flash_of_unstyled_content)을 방지하기 위해 파싱과 렌더링이 차단된다.

> `<script>` 역시 자바스크립트에 `document.write()` 와 같이 DOM을 변경시킬 수 있는 코드가 포함될 수 있기 때문에 파싱을 중단 시킨다.

Author: ```suggestion
`<script>`태그도 DOM을 변경하는 코드(`document.write()`)를 포함할 수 있기 때문에 파싱을 멈춘다.
```

파싱이 중단됨으로 인해 중요한 리소스 사용이 늦어지는 등 부작용이 생길 수 있는데, 브라우저에서는 이러한 문제를 완하시키기 위해 preload scanner를 사용해서 필요한 요청을 병렬적으로 처리할 수 있다.

### 2. Style

![style](./images/browser-rendering-performance/style.png)

DOM Tree파싱 이후 브라우저는 CSS를 파싱해서 세 가지 단계로 각 DOM노드의 스타일을 계산한다.

#### Step 1. CSS → 스타일 시트

![css_to_style_sheet](./images/browser-rendering-performance/css_to_style_sheet.png)

`<link>`태그로 로드하는 CSS, `<style>`, 그리고 inline style의 정보를 바탕으로 브라우저가 해석할 수 있는 스타일 시트를 생성한다.

#### Step 2. 단위 변환

- width: 50%
- padding: 2em 0
- font-size: 1rem

CSS는 px, %, em, rem등 다양한 단위로 작성할 수 있는데, rem등 상대적인 값은 픽셀로 치환되어 계산된다.

> '픽셀'로 표현하는 이유: 렌더링 마지막 단계에서 비트맵 데이터를 구성하는데, 비트맵 데이터가 픽셀로 구성되기 때문이다.

#### Step 3. 스타일 계산

![style_calc](./images/browser-rendering-performance/style_calc.png)

마지막으로, CSS 오버라이딩 등을 고려하여 요소의 최종 스타일을 계산한다.

### 3. Layout

![layout](./images/browser-rendering-performance/layout.png)

Layout단계에서는 DOM 트리와 스타일 시트를 기반으로 **어떤 요소를 어디에** 렌더링 해야하는 지에 대한 정보를 담고 있는 레이아웃 트리를 구성한다.

> 레이아웃 트리는 페이지에 렌더링 되는 정보만 포함하기 때문에 `display: none` 으로 처리된 요소는 포함되지 않는다.

Author: Layout 단계에서는 레이아웃 트리를 구성한다. DOM 트리와 스타일 시트를 기반으로 **어떤 요소**를 **어디에** 렌더링 해야하는 지를 결정한다. 이 레이아웃 트리엔 페이지에 렌더링 되는 정보만 포함하기 때문에 `display: none` 으로 처리된 요소는 포함되지 않는다.

![layout_cost](./images/browser-rendering-performance/layout_cost.gif)

레이아웃 트리를 구성하는 과정은 결코 간단하지 않다. 영상처럼 간단한 페이지에서도 폰트 크기에 따라 단락의 어느 부분에서 줄바꿈이 되어야 하는지 등을 모두 계산해야 하기 때문이다.

### 4. PrePaint

PrePaint단계는 레이어를 구성하기 위한 준비 단계로, 크게 두 가지 작업이 진행된다.

#### 1. Paint Invalidation

![paint_invalidation](./images/browser-rendering-performance/paint_invalidation.png)

> 이전 단계인 스타일, 레이아웃 단계 중 변화가 생겨 dirty bit이 생길 경우 이 단계에서 캐싱된 이전 paint기록을 무효화 한다.

Author: ```suggestion
이전 단계(스타일, 레이아웃 단계)에서 변화가 생길 경우(이를 dirty bit이라 한다.), 캐싱해둔 paint 기록을 무효화한다.
```

#### 2. Property Tree

![property_tree](./images/browser-rendering-performance/property_tree.png)

> Property Tree는 각 레이어에 할당되는 속성이다. 예를 들어 CSS속성 중 `transform`, `opacity`등의 속성을 적용할 경우 property tree에 반영되고 이후 레이어를 합치는 단계에서 필요한  효과를 빠르게 적용할 수 있다.

Author: ```suggestion
Property Tree는 각 레이어에 할당되는 속성이다. 예를 들어 CSS속성 중 `transform`, `opacity`등의 속성을 적용할 경우, property tree에 반영되고 이후 레이어를 합치는 단계에서 필요한 효과를 빠르게 적용할 수 있다.
```

문장이 잘 이해가 안 된다! 레이아웃 트리를 결정할 때, 적용되지 않고 프로퍼티 트리에 적용된다는 것인가?

기존에는 Property Tree에서 다뤄지는 데이터가 레이어에 함께 저장되어 있었기 때문에 특정 노드의 속성이 변경되면 해당 노드의 하위 노드에도 변경된 값을 반영하면서 노드를 순회해야 했다. 최신 Blink엔진에서는 이런 속성을 별도로 관리하고 각 노드에서는 Property Tree의 노드를 참조하는 방식으로 변경되었다.

### 5. Paint

> paint과정은 실제로 화면을 그리는 과정이 아니라 어떻게 그려야하는지에 대한 정보를 담고 있는 `Paint Records`를 생성하는 과정이다. 아래 세 가지 정보가 포함된다.

Author: ```suggestion
paint 과정은 실제로 화면을 그리는 과정이 아니라 **어떻게 그려야하는지**에 대한 정보를 담고 있는 `Paint Records`를 생성하는 과정이다. 이 레코드에는 아래 세 가지 정보가 포함된다.
```

- Action (e.g. Draw Rect)
- Position (e.g. 0, 0, 300, 300)
- Style (e.g. backgroundColor: red)

### 6. Layerize

![composition_forest](./images/browser-rendering-performance/composition_forest.png)

Layerize과정은 paint과정의 결과물을 사용해서 Composited Layer List라는 데이터를 생성하는 단계이다. layout단계에서 Layout Object로 구성된 Layout Tree가 생성되고 Layout Object에서 아래 조건을 만족하면 별도의 Paint Layer가 생성된다.

- 최상위 요소(root element)
- `position: relative, absolute` 사용
- 3D(`translate3d`, `preserve-3d`, ,..)나 perspective transform 사용
- `<video>`, `<canvas>` 태그 사용
- CSS `filter`나 alpha mask 사용

> 이 조건을 만족하지 않아 별도의 Paint Layer로 생성되지 않은 Layout Object는 가까운 상위 Pain Layer와 대응된다. (두 개 이상의 Layout Object가 하나의 Paint Layer로 다뤄질 수 있다.)

Author: ```suggestion
이 조건을 만족하지 않아 별도의 Paint Layer로 생성되지 않은 Layout Object는 가까운 상위 Paint Layer와 대응된다. (두 개 이상의 Layout Object가 하나의 Paint Layer로 다뤄질 수 있다.)
```
